compiler/rustc_expand/src/config.rs

```diff
@@ -394,26 +400,43 @@ impl<'a> StripUnconfigured<'a> {
 
     /// Determines if a node with the given attributes should be included in this configuration.
     fn in_cfg(&self, attrs: &[Attribute]) -> bool {
-        attrs.iter().all(|attr| !is_cfg(attr) || self.cfg_true(attr).0)
+        attrs.iter().all(|attr| !is_cfg(attr) || self.cfg_true(attr, self.lint_node_id).as_bool())
     }
 
-    pub(crate) fn cfg_true(&self, attr: &Attribute) -> (bool, Option<MetaItem>) {
-        let meta_item = match validate_attr::parse_meta(&self.sess.psess, attr) {
-            Ok(meta_item) => meta_item,
+    pub(crate) fn cfg_true(&self, attr: &Attribute, node: NodeId) -> EvalConfigResult {
+        // We need to run this to do basic validation of the attribute, such as that lits are valid, etc
+        // FIXME(jdonszelmann) this should not be necessary in the future
+        match validate_attr::parse_meta(&self.sess.psess, attr) {
+            Ok(_) => {}
             Err(err) => {
                 err.emit();
-                return (true, None);
+                return EvalConfigResult::True;
             }
-        };
+        }
+
+        // Unsafety check needs to be done explicitly here because this attribute will be removed before the normal check
+        deny_builtin_meta_unsafety(
+            self.sess.dcx(),
+            attr.get_normal_item().unsafety,
+            &rustc_ast::Path::from_ident(attr.ident().unwrap()),
+        );
 
-        validate_attr::deny_builtin_meta_unsafety(&self.sess.psess, &meta_item);
+        let Some(rustc_hir::Attribute::Parsed(AttributeKind::Cfg(cfg, _))) =
```

Comment:
So I don't quite get why we're still calling into the full attribute parse system here. In theory, cfg doesn't need a defined parser impl since it's *never* parsed late. That's unlike repr which is parsed both early and late and we use parse limited to trigger the existimg parser early. In theory we could just create a Context and call a function in rustc_attr_parsing which instead of returning an AttributeKind we now need to match on again, just returns the CfgEntry. Doesn't even have to implement AttributeParser, we just want it to use the structs from rustc_attr_parsing/src/parser.rs so we can also call into that code from "real" parsers where the same syntax is used. So just a standalone function that takes an AttrArgs and returns a CfgEntry

Reply:
A bit like parse_cfg but in the new system